### PR TITLE
Refactor and make cache usable

### PR DESCRIPTION
**Pull request by rylandg**
This is mostly just an iteration PR for feedback, don't merge.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -14,14 +14,9 @@ RUN apt-get update && apt-get install -y python-dev python-pip curl && apt-get b
...
-ADD .binaris/.pycache /cache
-COPY docker/* /function/
-WORKDIR /function/
-
-# Re-use PIP cache between runs.  Do *NOT* copy this out!
-ENV XDG_CACHE_HOME=/cache/
-
-CMD [ "./install.sh" ]
+RUN pip install virtualenv
+COPY oldrequirements.txt /tmp/oldrequirements.txt
```
Comment (arielshaqed) [inline Dockerfile:18]: Please keep in the same directory (e.g. `requirements.txt`, `requirements.txt.bk`, `requirements.txt.bk2`).  I don't care about the names, I don't particularly care how many you keep of each file; I _do_ care they're all in the same directory.

Also the use of `/tmp` as a working directory is problematic.  I'd prefer some (any) other directory.

File: build.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+else
+    DOCKER='sudo docker'
+fi
+
+tmpdir='/tmp/bobdeps'
+
+mkdir -p $tmpdir
+cp $dir/Dockerfile $tmpdir
+
+if [ -f $PWD/oldrequirements.txt ]; then
```
Comment (arielshaqed) [inline build.sh:21]: Why can't all these copies be in the Docker file?  (If you must, pass as variables into the file, or template them out.)

**Review by arielshaqed — APPROVED** [review]
- Why are we creating `requirements.txt`?  Why are we creating an `oldrequirements.txt`??

## MUCH nicer!

**Comment by rylandg** [comment]
@arielshaqed there is a fundamental misunderstanding about what these changes do. 

It will help if I answer your question regarding `requirements.txt` and `oldrequirements.txt` first. The purpose of `oldrequirements.txt` is very simple, it allows pip to utilize the `cache` for dependencies in 9/10 cases for the user. 

Imagine that you have a long list of dependencies, during the lifetime of a given service dependencies will be added/removed but generally 80-90% of them will remain static. With a naive approach every change in `requirements.txt` would trigger a complete `cache` deletion (both in Docker & pip terms since pip resides within a layer of Docker). This means adding a single dependency will take (prev install/download time + added module download/install time).

But as we noted before 80-90% of the dependencies won't change during the lifetime of an app. A clever solution to this problem is a stable state of sorts. If we 'freeze' the initial `requirements.txt` and save it into `oldrequirements.txt` when the Docker step runs it will use the local `pip` cache created in the previous layer (where we do `venv` with `oldrequirements.txt`) instead of downloading every package from the registry. This means the max number of package fetches from `pip` is limited to the size of the `difference set` of old vs current requirements. 

I hope that clarifies how it works from a functional standpoint. Now I want to briefly discuss the overall design. 

I asked on our recent call about expectations. The understanding we reached (as far as I remember)  was that this tool only needs to be a glorified `requirements.txt` compiler that allows local dependencies to run correctly on a remote bolt state. My extrapolation was that `bob` does not need to be a generic/usable docker image but rather a single purpose tool for building working binaries. This is why I'm confused about issues regarding directories (`tmp` specifically) within the Docker image itself. As I see it this is the expected user flow.

1. User creates a simple `bn` function with some python dependencies
2. User decides that the majority of dependencies are stable and `cp` his `requirements.txt` file (note these files reside in the directory with his function/binaris.yml) into `oldrequirements.txt`.
3. User runs `bob` from that same directory and `bob` first installs the `oldrequirements.txt` and then `requirements.txt` on top of that.
4. User deploys function without any other actions and the expected dependencies are in `dist` folder accessible by `bolt`.
4. User may make changes to `requirements.txt` and reruns `bob`, the only packages that are fetched are the ones not present in his stable `oldrequirements.txt`.
5. At some point the user decides the delta between old and current is too much and repeats *step 2*

I hope this clarifies things substantially. The only other comment I would like to make is that I carried over the practice of a `safe and sane default` that was in the original `bob`. Therefore even with no local `oldrequirements.txt` the versioned `requirements.txt` is used for caching.

**Pull request merged**